hal/src/thumbv6m/sercom/v2/uart.rs

```diff
@@ -0,0 +1,1937 @@
+    // Once the list of [`PadNum`]s is empty, we're done with this element.
+    ( ( $($Perm:ident)* ) [ ] [ $($Pads:ident)+ ] ) => { };
+}
+
+padnum_permutations!( () [NoneT Pad0 Pad1 Pad2 Pad3] );
+
+//=============================================================================
+// Pads
+//=============================================================================
+
+/// Container for a set of SERCOM [`Pad`]s
+///
+/// A [`Sercom`] can use up to four [`Pin`]s as peripheral [`Pad`]s, but only
+/// certain `Pin` combinations are acceptable. In particular, all `Pin`s must be
+/// mapped to the same `Sercom` (see the datasheet). This HAL makes it
+/// impossible to use invalid `Pin`/`Pad` combinations, and the [`Pads`] struct
+/// is responsible for enforcing these constraints.
+///
+/// A `Pads` type takes up to five type parameters. The first specifies the
+/// `Sercom`. The remaining four, `RX`, `TX`, `RTS` and CTS, represent the Tx
+/// Rx, Ready to send and Clear to send `Pad`s respectively, and they default to
+/// [`NoneT`]. These type parameters take two different forms, depending on the
+/// chip. For SAMD21 chips, they are effectively [`OptionalPinId`]s. While for
+/// SAMD11 chips, they are optional ([`PadNum`], [`PinId`]) tuples. See the
+/// [`GetPad`] trait for an explanation of the reasoning here.
+///
+/// ```
+/// use atsamd_hal::gpio::v2::{PA04, PA05, PA08, PA09};
+/// use atsamd_hal::sercom::v2::{Sercom0, uart};
+/// use atsamd_hal::sercom::v2::pad::{Pad0, Pad1};
+/// use atsamd_hal::typelevel::NoneT;
+///
+/// // For SAMD21 chips
+/// type Pads = uart::Pads<Sercom0, PA08, NoneT, PA09>;
+///
+/// // For SAMD11 chips
+/// type Pads = uart::Pads<Sercom0, (Pad0, PA04), NoneT, (Pad1, PA05)>;
+/// ```
+///
+/// `Pads` are created using the builder pattern. Start by creating an empty set
+/// of `Pads` using [`Default`]. Then pass each respective `Pin` using the
+/// corresponding methods. Both `v1::Pin` and `v2::Pin` types are accepted here.
+///
+/// To be accepted as part of a [`ValidConfig`], a set of `Pads` must do two
+/// things: specify a type for at least one of `RX` or `TX`; and
+/// satisfy the [`RxpoTxpo`] trait.
+///
+/// ```
+/// use atsamd_hal::target_device::Peripherals;
+/// use atsamd_hal::gpio::v2::Pins;
+/// use atsamd_hal::sercom::v2::{Sercom0, uart};
+///
+/// let mut peripherals = Peripherals::take().unwrap();
+/// let pins = Pins::new(peripherals.PORT);
+/// let pads = uart::Pads::<Sercom0>::default()
+///     .rx(pins.pa08)
+///     .tx(pins.pa10);
+/// ```
+///
+/// [`Pin`]: crate::gpio::v2::pin::Pin
+/// [`PinId`]: crate::gpio::v2::pin::PinId
+/// [`OptionalPinId`]: crate::gpio::v2::pin::OptionalPinId
+pub struct Pads<S, RX = NoneT, TX = NoneT, RTS = NoneT, CTS = NoneT>
+where
+    S: Sercom,
+    RX: GetOptionalPad<S>,
+    TX: GetOptionalPad<S>,
+    RTS: GetOptionalPad<S>,
+    CTS: GetOptionalPad<S>,
+{
+    rx: RX::Pad,
+    tx: TX::Pad,
+    ready_to_send: RTS::Pad,
+    clear_to_send: CTS::Pad,
```

Comment:
Minor nit: to me it feels a little inconsistent to abbreviate rx and tx, but not rts and cts.  I've got a very weak preference for receive & transmit rather than rts & cts, just because they're easier to search for, but either would be fine.

Reply:
Fair enough. Though I prefer keeping tx and rx, only because `receive` and `transmit` could be mistaken as the act of actually receiving and transmitting data, rather than just assigning pins to pads.

Reply:
I agree with @ianrrees. But also keep in mind that this isn't user-facing.

Reply:
Ah I see. I'll change the struct fields to receive/transmit, but change the user-facing methods to rts and cts. I hope this isn't too confusing?